libsolidity/experimental/analysis/TypeInference.cpp

```diff
@@ -708,11 +716,22 @@ bool TypeInference::visit(TypeClassInstantiation const& _typeClassInstantiation)
 			m_errorReporter.typeError(6948_error, _typeClassInstantiation.location(), "Missing function: " + name);
 			continue;
 		}
+		Type instantiatedClassFunctionType = TypeEnvironmentHelpers{*m_env}.substitute(classFunctionType, classVar, type);
+
 		Type instanceFunctionType = functionTypes.at(name);
 		functionTypes.erase(name);
 
-		if (!newEnv.typeEquals(instanceFunctionType, classFunctionType))
-			m_errorReporter.typeError(7428_error, _typeClassInstantiation.location(), "Type mismatch for function " + name + " " + TypeEnvironmentHelpers{newEnv}.typeToString(instanceFunctionType) + " != " + TypeEnvironmentHelpers{newEnv}.typeToString(classFunctionType));
+		if (!m_env->typeEquals(instanceFunctionType, instantiatedClassFunctionType))
+			m_errorReporter.typeError(
+				7428_error,
+				_typeClassInstantiation.location(),
+				fmt::format(
+					"Instantiation function '{}' does not match the declaration in the type class ({} != {}).",
```

Comment:
Shouldn't we have some test to cover this error here? Maybe something like this:
```solidity
pragma experimental solidity;

type bool = __builtin("bool");
type word = __builtin("word");

class Self: C {
    function f(self: Self) -> bool;
}

instantiation word: C {
    function f(self: bool) -> bool {}
}
```
I guess this should already trigger this error, since the instantiation of `f` would be `bool -> bool` and not  `word -> bool`.

Reply:
Generally, I'd not go overboard with testing this stuff just yet (it's experimental after all) - but yeah, in this case it's easy to do and this won't change in the future, so a simple test makes sense.

Reply:
Test added. And yeah, I intentionally did not try to go for full coverage here. Many corner cases are not even testable due to ICEs so this is not feasible at this point.